storage: don't use quota pool throttling for NodeLiveness range

nvb opened the pull request:
We don't want to throttle updates to the `NodeLiveness` range even
if a follower is falling behind because this could result in
cascading failures.

I suspected that this may have caused issues see previously with https://blog.stathat.com/2014/12/22/fix_ec2_network_issue_skb_rides_the_rocket.html,
although that claim remains unsubstantiated as I haven't been able to find any
reference to the NodeLiveness range being throttled. Still, this seems like a
good idea.

nvb committed:
storage: clean up quota pool logic for active replicas

Follow up to #19819.

petermattis commented:
A few drive-by comments. Someone else should look at this more thoroughly.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1002 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--GbXzWxn_CAGycKyD:-L--GbXzWxn_CAGycKyE:b-qhmylo) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1002)):*
> ```Go
> 		return false
> 	}
> 	if bytes.Equal(r.mu.state.Desc.StartKey, keys.NodeLivenessPrefix) {
> ```

Should this be `r.ContainsKeyRange(keys.NodeLivenssSpan)`? That mirrors the check in `MaybeGossipNodeLiveness()`.

---

*[pkg/storage/replica.go, line 1018 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--HIoGsHv6lOHqKG2l:-L--HIoGsHv6lOHqKG2m:b-z0pt5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1018)):*
> ```Go
> 	defer r.mu.Unlock()
> 
> 	if !r.quotaPoolEnabledLocked() {
> ```

Is it worthwhile to move the node-liveness check out of the fast-path? For example, you could check only before `proposalQuota` is created and before the assertion that `proposalQuota` is not nil on the leader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-:-L--HgV-TAgph2E-K1-4:b-jh0fxi)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica.go, line 1002 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--GbXzWxn_CAGycKyD:-L-DHN2L0tUWYfF87yfx:bo8hov0) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1002)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Should this be `r.ContainsKeyRange(keys.NodeLivenssSpan)`? That mirrors the check in `MaybeGossipNodeLiveness()`.
</blockquote></details>

We may have to allow splitting of node liveness somewhere down the road and it'd be nice if the code didn't have to change in that event as it's not unlikely to be overlooked. How about checking whether `StartKey` has `keys.NodeLivenessPrefix` as a prefix?

Don't feel too strongly.

---

*[pkg/storage/replica.go, line 1018 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--HIoGsHv6lOHqKG2l:-L-DIzUThqj3CaE_Hnoh:b-a4i0o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1018)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is it worthwhile to move the node-liveness check out of the fast-path? For example, you could check only before `proposalQuota` is created and before the assertion that `proposalQuota` is not nil on the leader.
</blockquote></details>

Unless you really think this matters for performance (usually the mismatch in the comparison is on the first byte) I'd avoid that for now (we may want to cherry-pick this?). In fact, I'd even have expected to leave the quota pool machinery in its place, just to skip acquiring the quota, to make sure the liveness range works 99.9% like the others.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-:-L-DJm38obx11MnzEZ-z:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: don't use quota pool throttling for NodeLiveness range

We don't want to throttle updates to the `NodeLiveness` range even
if a follower is falling behind because this could result in
cascading failures.

nvanbenschoten commented:
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/storage/replica.go, line 1002 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--GbXzWxn_CAGycKyD:-L-8eD-TiduvfSUHV-K2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1002)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

We may have to allow splitting of node liveness somewhere down the road and it'd be nice if the code didn't have to change in that event as it's not unlikely to be overlooked. How about checking whether `StartKey` has `keys.NodeLivenessPrefix` as a prefix?

Don't feel too strongly.
</blockquote></details>

Done.

---

*[pkg/storage/replica.go, line 1018 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288#-L--HIoGsHv6lOHqKG2l:-L-DX2lZ1Dm8HZwLMmHA:b-x4vu00) ([raw file](https://github.com/cockroachdb/cockroach/blob/479be14753c981d3bbeac32368d6744490179c9f/pkg/storage/replica.go#L1018)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Unless you really think this matters for performance (usually the mismatch in the comparison is on the first byte) I'd avoid that for now (we may want to cherry-pick this?). In fact, I'd even have expected to leave the quota pool machinery in its place, just to skip acquiring the quota, to make sure the liveness range works 99.9% like the others.
</blockquote></details>

I followed @tschottdorf's suggestion and made the change in `maybeAcquireProposalQuota` instead. This seems like a much better place for this logic as it minimizes the differences between the `NodeLiveness` range and the other ranges. It also simplifies logic on startup as ranges split at the fixed split points to create a discrete `NodeLiveness` range.

I also agree that having the check on the fast-path shouldn't make much of a difference as this should almost always be a single-byte comparison.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20288)*
<!-- Sent from Reviewable.io -->
